### PR TITLE
EdgeAgent: Fix starting of stopped modules

### PR DESCRIPTION
**Pull request by varunpuranik**
- If a module is in a "stopped" state but should be running, start it. 
- If a module's desired status changes, then don't redeploy the module, only start or stop it.

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/Diff.cs [hunk 1]
```diff
@@ -8,56 +8,45 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Core
 
     public class Diff
     {
-        public Diff(IList<IModule> updated, IList<string> removed)
+        public Diff(IList<IModule> added, IList<IModule> updated, IList<IModule> desiredStatusUpdated, IList<string> removed)
```
Comment (myagley) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/Diff.cs]: We should add a comment that discussed what each of these `IList<IModule>`s is used for.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/ModuleSet.cs [hunk 2]
```diff
@@ -38,17 +38,6 @@ public ModuleSet(IImmutableDictionary<string, IModule> modules)
 
         public bool TryGetModule(string key, out IModule module) => this.Modules.TryGetValue(key, out module);
 
-        public ModuleSet ApplyDiff(Diff diff)
```
Comment (myagley) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/ModuleSet.cs]: Does this method no longer make sense?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/planners/HealthRestartPlanner.cs [hunk 3]
```diff
@@ -11,13 +11,17 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Core.Planners
     using Microsoft.Azure.Devices.Edge.Util;
     using Microsoft.Extensions.Logging;
     using Newtonsoft.Json;
+    using Nito.AsyncEx;
```
Comment (myagley) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/planners/HealthRestartPlanner.cs]: How are we going to test these changes? I see we've added a bunch of unit tests. It sounds like the arm32 E2E will also catch some of these?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/ModuleSet.cs (hunk 2 above)
Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/ModuleSet.cs]: No, this was used originally to apply the deployment update from the twin patch. That is done in EdgeAgentConnection now, with DeploymentConfig.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/Diff.cs (hunk 1 above)
Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/Diff.cs]: Sure, will do.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/planners/HealthRestartPlanner.cs (hunk 3 above)
Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/planners/HealthRestartPlanner.cs]: Yeah, I added unit tests. And yes, the arm32 e2e tests will validate the fix as well. Apparently we are seeing this quite frequently now.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/Diff.cs (hunk 1 above)
Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/Diff.cs]: Added comments on the properties..

**Pull request merged**